### PR TITLE
Re-attempt to delete temporary upload if the file is locked by another process

### LINKED ISSUE #12339: Error when uploading images
- Gitea version (or commit ref): 1.12.2 built with GNU Make 4.1, go1.14.4 : bindata, sqlite, sqlite_unlock_notify
- Git version: 2.27.0
- Operating system: Windows Server 2019
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x ] No
  - [ ] Not relevant
- Log gist:

## Description
Every time I upload an image I get this error. Weird thing is it uploads the file fine and I can use it. I'm guessing its having issues deleting this temp file. Not sure why since I have not had any issues creating repositories on this server or pushing updates.

image

...

### PR DESCRIPTION
**Pull request by zeripath**
Fix #12339

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: models/upload.go [hunk 1]
```diff
@@ -130,6 +132,12 @@ func DeleteUploads(uploads ...*Upload) (err error) {
 		}
 
 		if err := os.Remove(localPath); err != nil {
+			i := 0
+			for i < 5 && strings.Contains(err.Error(), "The process cannot access the file because it is being used by another process") {
```
Comment (silverwind) [inline models/upload.go]: I imagine this string will be affected by the localization setting of Windows. Are there error codes available?

**Comment by silverwind** [comment]
File deletion on Windows can be a pain with the locking. I wonder if there are any third-party golang modules similar to https://github.com/isaacs/rimraf/blob/master/rimraf.js that deal with the various error scenarios.

File: models/upload.go (hunk 1 above)
Comment (zeripath) [reply models/upload.go]: If you can find one... I couldn't

Comment (silverwind) [reply models/upload.go]: Maybe just retry regardless of error message? Could possibly abstract that into a utility function like `RemoveFileGraceful`.

(Does golang generally really not provide error codes, only messages?)

**Comment by zeripath** [comment]
Fine. Let's do this properly - replace all uses of os.Remove / os.RemoveAll with util.Remove / util.RemoveAll which will reattempt on certain errors.

File: modules/util/remove.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+package util
+
+import (
+	"os"
+	"syscall"
+	"time"
+)
+
+// Remove removes the named file or (empty) directory with at most 5 attempts.Remove
```
Comment (silverwind) [inline modules/util/remove.go]:
```suggestion
// Remove removes the named file or (empty) directory with at most 5 attempts.
```

File: modules/util/remove.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+			break
+		}
+		unwrapped := err.(*os.PathError).Err
+		if unwrapped == syscall.EBUSY || unwrapped == syscall.ENOTEMPTY || unwrapped == syscall.EPERM {
+			// try again
+			<-time.After(100 * time.Millisecond)
+			continue
+		}
+
+		// TODO: Consider handling EMFILE?
```
Comment (silverwind) [inline modules/util/remove.go]: There is a slightly more up to date rimraf implementation at https://github.com/nodejs/node/blob/master/lib/internal/fs/rimraf.js which defineds `'EBUSY', 'EMFILE', 'ENFILE', 'ENOTEMPTY', 'EPERM'` as retryable. Maybe we should just copy that.

File: models/upload.go [hunk 4]
```diff
@@ -129,7 +132,13 @@ func DeleteUploads(uploads ...*Upload) (err error) {
...
 		}
 
-		if err := os.Remove(localPath); err != nil {
+		if err := util.Remove(localPath); err != nil {
+			i := 0
+			for i < 5 && strings.Contains(err.Error(), "The process cannot access the file because it is being used by another process") {
+				<-time.After(100 * time.Millisecond)
+				err = util.Remove(localPath)
+				i++
+			}
```
Comment (zeripath) [inline models/upload.go:141]:
```suggestion
		if err := util.Remove(localPath); err != nil {
```

Comment (zeripath) [inline models/upload.go]:
```suggestion
```

File: models/upload.go [hunk 5]
```diff
@@ -11,8 +11,11 @@ import (
 	"mime/multipart"
 	"os"
 	"path"
+	"strings"
+	"time"
```
Comment (zeripath) [inline models/upload.go]:
```suggestion
```

File: modules/util/remove.go (hunk 3 above)
Comment (zeripath) [reply modules/util/remove.go]: EMFILE and ENFILE only occur on file open - can they really occur on deletion?

Comment (silverwind) [reply modules/util/remove.go]: If we only unlink single files without any `stat` calls, it should be safe. Just for recursive deletion that may involve readdir, they can happen.

File: models/unit_tests.go [hunk 6]
```diff
@@ -68,18 +69,18 @@ func MainTest(m *testing.M, pathToGiteaRoot string) {
 	}
 
 	if err = removeAllWithRetry(setting.RepoRootPath); err != nil {
```
Comment (techknowlogick) [inline models/unit_tests.go]: the three calls in this file to `removeAllWithRetry` could be replaced with `util.RemoveAll` (or at very least remove loop in `removeAllWithRetry` and have just one call to `util.RemoveAll`), otherwise it'll be 5 * 20 attempts to rm the files

File: go.sum [hunk 7]
```diff
@@ -711,8 +711,6 @@ github.com/yuin/goldmark v1.1.7/go.mod h1:3hX8gzYuyVAZsxl0MRgGTJEmQBFcNTphYh9dec
 github.com/yuin/goldmark v1.1.22/go.mod h1:3hX8gzYuyVAZsxl0MRgGTJEmQBFcNTphYh9decYSb74=
 github.com/yuin/goldmark v1.1.25 h1:isv+Q6HQAmmL2Ofcmg8QauBmDPlUUnSoNhEcC940Rds=
 github.com/yuin/goldmark v1.1.25/go.mod h1:3hX8gzYuyVAZsxl0MRgGTJEmQBFcNTphYh9decYSb74=
-github.com/yuin/goldmark v1.1.32 h1:5tjfNdR2ki3yYQ842+eX2sQHeiwpKJ0RnHO4IYOc4V8=
-github.com/yuin/goldmark v1.1.32/go.mod h1:3hX8gzYuyVAZsxl0MRgGTJEmQBFcNTphYh9decYSb74=
```
Comment (silverwind) [inline go.sum]: unrelated?

File: modules/util/remove.go (hunk 3 above)
Comment (silverwind) [reply modules/util/remove.go]: I think we should retry EMFILE and ENFILE for the `RemoveAll` case which does involve `stat`.

File: go.sum (hunk 7 above)
Comment (zeripath) [reply go.sum]: yes, the go.sum needs to be updated at some point.

However, fine I'll remove it

File: models/unit_tests.go (hunk 6 above)
Comment (zeripath) [reply models/unit_tests.go]: done

File: modules/util/remove.go (hunk 3 above)
Comment (zeripath) [reply modules/util/remove.go]: Fine

File: go.sum (hunk 7 above)
Comment (silverwind) [reply go.sum]: Probably best done in a full vendor update PR.

**Review by silverwind — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]
